pkgs/build-support/setup-hooks/make-binary-wrapper.sh

```diff
@@ -0,0 +1,274 @@
+# Assert that FILE exists and is executable
+#
+# assertExecutable FILE
+assertExecutable() {
+    local file="$1"
+    [[ -f "$file" && -x "$file" ]] || \
+        die "Cannot wrap '$file' because it is not an executable file"
+}
+
+# Generate a binary executable wrapper for wrapping an executable.
+# The binary is compiled from generated C-code using gcc.
+# makeBinaryWrapper EXECUTABLE OUT_PATH ARGS
+
+# ARGS:
+# --argv0       NAME    : set name of executed process to NAME
+#                         (otherwise it’s called …-wrapped)
+# --set         VAR VAL : add VAR with value VAL to the executable’s
+#                         environment
+# --set-default VAR VAL : like --set, but only adds VAR if not already set in
+#                         the environment
+# --unset       VAR     : remove VAR from the environment
+# --add-flags   FLAGS   : add FLAGS to invocation of executable
+
+# --prefix          ENV SEP VAL   : suffix/prefix ENV with VAL, separated by SEP
+# --suffix
+
+# To troubleshoot a binary wrapper after you compiled it,
+# use the `strings` command or open the binary file in a text editor.
+makeBinaryWrapper() {
+    assertExecutable "$1"
+    makeDocumentedCWrapper "$1" "${@:3}" | gcc -Os -x c -o "$2" -
+}
+
+# Syntax: wrapProgramBinary <PROGRAM> <MAKE-WRAPPER FLAGS...>
+wrapProgramBinary() {
+    local prog="$1"
+    local hidden
+
+    assertExecutable "$prog"
+
+    hidden="$(dirname "$prog")/.$(basename "$prog")"-wrapped
+    while [ -e "$hidden" ]; do
+      hidden="${hidden}_"
+    done
+    mv "$prog" "$hidden"
+    # Silence warning about unexpanded $0:
+    # shellcheck disable=SC2016
+    makeBinaryWrapper "$hidden" "$prog" --argv0 '$0' "${@:2}"
```

Comment:
```suggestion
    makeBinaryWrapper "$hidden" "$prog" --argv0 "$0" "${@:2}"
```
otherwise we get erroneous behaviour like this:
```
execve("/nix/store/sz6zhvzifmh8mj0frqmn0zmza2kzcirb-foo/bin/.foo-wrapped", ["$0"], 0x13712a0 /* 76 vars */) = 0
```

could you add a test for the generated executables as well?

Reply:
ah, now i see where this come from: seems it was lifted straight from `makeWrapper`, and thereby lost a level of expansion that would've happened in the generated shell script. the generated binary of course won't do this, causing this behaviour

Reply:
One option here could be to make `'$0'` a "magic value" when passed to --argv0 (making us not override argv[0]). That way, if existing scripts use this trick elsewhere with makeWrapper, they will continue to work with makeBinaryWrapper.

Reply:
The downside is that people might expect `'$1'` or `'$SOME_VAR'` to work as well, or try to pass `'$0'` to other things than --argv0

Reply:
the trick is that this `$0` ends up in the generated wrapper like so:
```
exec -a "$0" "/nix/store/rz6w8y7aci147cy3i2hiw5m7v9nyssm2-foo/bin/.foo-wrapped"  "$@"
```
would argue that all other uses of this trick (eg `--set FOO '$BAR'`) are highly suspect, and adding special casing to enable one specific instance seems like a detriment rather than an improvement. not modifying argv[0] would be fine, but it should be predicated on `--argv0` not being given instead of on the value it receives

Reply:
If you don't pass `--argv0` to the existing makeWrapper, it will be set to the name of the wrapped executable. Having makeBinaryWrapper behave differently from makeWrapper here might not be ideal. An alternative to allowing a magic value '$0' is to add another flag. Something like `--inherit-argv0` which takes 0 extra arguments.

Reply:
`--inherit-argv0` sounds good. :+1: